improve the branches model to be more git-like (#2994)
brimdata/zed: docs/lake/design.md

The create command initiates a new pool with a single branch called `main`.

> Zed lakes can be used without branches.  When referencing a pool without
> a branch, the tooling presumes the "main" branch as the default, and everything
> can be done on main without having to think about branching.

### Branch

> A branch is simply a named pointer to a commit object in the zed lake.

Reviewer: ```suggestion
A branch is simply a named pointer to a commit object in the Zed lake.
```

The rest of the doc is consistent about capitalizing "Zed" when talking about the project and `zed` when mentioning the command line tool.

Like git, commit objects are arranged into a tree and represent the entire
commit history of the lake.
A branch is created with the `branch` command, e.g.,
```
zed lake branch -p logs@main staging
```
This creates a new branch called "staging" in pool "logs", which points to

> same commit object as the "main" branch.  Commits to the "staging" branch

Reviewer: ```suggestion
the same commit object as the "main" branch.  Commits to the "staging" branch
```

will be added to the commit history without affecting the "main" branch

> and each branch can be queried indepedently at any time.

Reviewer: ```suggestion
and each branch can be queried independently at any time.
```

a "table" as all Zed data is _self describing_ and can be queried in a
schema-agnostic fashion.  Data of any _shape_ can be stored in any pool
and arbitrary data _shapes_ can coexist side by side.

#### Data Segmentation

In a `load` operation, a commit is broken out into units called _data objects_
where a target object size is configured into the pool,
typically 100-500MB.  The records within each object are sorted by the pool key(s).
A data object is presumed by the implementation
to fit into the memory of an intake worker node
so that such a sort can be trivially accomplished.

Data added to a pool can arrive in any order with respect to the pool key.
While each object is sorted before it is written,
the collection of objects is generally not sorted.

### Log

Like `git log`, the command `zed lake log` prints a history of commit objects
starting from any commit.  The log can be displayed with the `log` command,
e.g.,
```
zed lake log -p pool@branch
```
To understand the log contents, the `load` operation is actually
decomposed into two steps under the covers:
an `add` step stores one or more
new immutable data objects in the lake and a `commit` operation

> materialize the objects into a branch with an ACID transaction.

Reviewer: ```suggestion
materializes the objects into a branch with an ACID transaction.
```

This updates the branch pointer to point at a new commit object
referencing the data objects where the new commit object's parent
points at the branch's previous commit object, thus forming a path
through the object tree.

> Note that following the pointers of a sequence of commit objects
> each stored independently in cloud storage can have tremendously
> high latency.  Fortunately, all of these objects are immutable and
> any commit object thus has a predetermined state that can be computed
> from its predecessors and persisted as a snapshot and cached
> in memory (or in redis etc).

Every commit object and data object is named by and referenced
using globally unique [KSUIDs](https://github.com/segmentio/ksuid),
called a `commit ID` or a data `object ID`, respectively.

The log command prints the commit ID of each commit object in that path
from the current pointer back through history to the first commit object.

A commit object includes
an optional author and message, along with a required timestamp,

> that is stored in the commit journal for reference.  This values may

Reviewer: ```suggestion
that is stored in the commit journal for reference.  These values may
```

}
```
None of the fields are used by the Zed lake system for any purpose
except to provide information about the commit object to the end user
and/or end application.  Any ZSON/ZNG data can be stored in the `Data` field
allowing external applications to implement arbitrary data provenance and audit
capabilities by embedding custom metadata in the commit journal.

> The Data field is not yet implemented.

### Merge

Data is merged from one branch into another with the `merge` command, e.g.,
```
zed lake merge -p logs@updates main
```
where the "updates" branch is being merged into the "main" branch.

A merge operation finds a common ancestor in the commit history then

> computes the set of changes needed for target branch to reflect the

Reviewer: ```suggestion
computes the set of changes needed for the target branch to reflect the
```

The `commit` operation is _transactional_.  This means that a query scanning
a pool sees its entire data scan as a fixed "snapshot" with respect to the
commit history.  In fact, the Zed language allows a commit object (created
at any point in the past) to be specified with the `@` suffix to a
pool reference, e.g.,
```
zed lake query -z 'from logs@1tRxi7zjT7oKxCBwwZ0rbaiLRxb | count() by field'
```
In this way, a query can time-travel through the journal.  As long as the
underlying data has not been deleted, arbitrarily old snapshots of the Zed
lake can be easily queried.

If a writer commits data after a reader starts scanning, then the reader
does not see the new data since it's scanning the snapshot that existed
before these new writes occurred.

Also, arbitrary metadata can be committed to the log as described below,
e.g., to associate index objects or derived analytics to a specific
journal commit point potentially across different data pools in
a transactionally consistent fashion.

#### Time Travel

> While time-travel through commit history provides one means to explore

Reviewer 2: ```suggestion
While time travel through commit history provides one means to explore
```

does not see the new data since it's scanning the snapshot that existed

> before these new writes occur.

Reviewer: ```suggestion
before these new writes occurred.
```

is not relied on for causal relationships since clock skew can violate
such an assumption).

> Data objects are referred to by zero or more commit objects, where the =

Reviewer: ```suggestion
Data objects are referred to by zero or more commit objects, where the
```

> commits are maintained referenced by the branches journal described below.

Reviewer: I'm not sure what suggestion to make to fix this sentence, but "maintained referenced by" reads funny.

Author: deleting

> are encoded into its metadata section so there is no need to have a separate
> seek index for the columnar object.

#### Commit History

A pool's commit history is the definitive record of the evolution of data in
that pool in a transactionally consistent fashion.

Each commit object entry is identified with its `commit ID`.
Objects are immutable and uniquely named so there is never a concurrent write
condition.

The `add` and `commit` operations are transactionally stored

> in a chain of commit object.  Any number of adds (and deletes) may appear

Reviewer: ```suggestion
in a chain of commit objects.  Any number of adds (and deletes) may appear
```

in a commit object.  All of the operations that belong to a commit are
identified with a commit identifier (ID).

As each commit object points to its parent (except for the initial commit

> in main), the collection commit objects in a pool forms a tree.

Reviewer: ```suggestion
in main), the collection of commit objects in a pool forms a tree.
```

Each commit object contains a sequence of _actions_:

* `Add` to add a data object reference to a pool,
* `Delete` to delete a data object reference from a pool,
* `AddIndex` to bind an index object to a data object to prune the data object
from a scan when possible using the index,
* `DeleteIndex` to remove an index object reference to its data object, and
* `Commit` for providing metadata about each commit.

The actions are not grouped directly by their commit tag but instead each
action embeds the KSUID of its commit tag.

By default, `zed lake log` outputs an abbreviated form of the log as text to
stdout, similar to the output of `git log`.

However, the log represents the definitive record of a pool's present
and historical content, and accessing its complete detail can provide
insights about data layout, provenance, history, and so forth.  Thus,
Zed lake provides a means to query a pool's configuration state as well,
thereby allowing past versions of the complete pool and branch configurations
as well as all of their underlying data to be subject to time travel.

> To interrogate the underlying trasaction history of the branches and

Reviewer: ```suggestion
To interrogate the underlying transaction history of the branches and
```

* transaction journal incomplete
* no recursive journal pool
* no columnar support

## CLI tool naming conventions

> > This is a work in in progress.

Reviewer: ```suggestion
> This is a work in progress.
```

Now that we have a comprehensive branching model, the `-p` argument to the
`zed lake` commands needs to be revisited.

We need to have a consistent syntax to refer to pools, branches, and commits.
Some commands require branch names (e.g., branch, rename, merge, etc) while
others require a commit (e.g., log or a scan in the Zed `from` operator).

The proposal on the table is to refer to commits things and branch references
both as an `@` character following the pool being referenced, e.g.,
```
pool@commit
```
where commit is a KSUID of the commit object, or
```
pool@branch
```
where branch is a name (that cannot be a KSUID).

Confusion arises because a commit can be referred to by a branch name
and, depending on context, a branch can refer to a commit or to
a mutable reference of that branch.  For the technical discussion here, we will
refer to the branch/commit dichotomy as a `commitish` and the l-value
nature of a branch as a `ref`.

Both a `commitish` and a `ref` may drop omit `@` suffix
and refer solely to the pool name, in which case it becomes pool@main.

> Complicating matters further, a `ref` may be specified a simple name

Reviewer: ```suggestion
Complicating matters further, a `ref` may be specified via a simple name
```

In some contexts (e.g., `create` and `drop`),
only a pool name makes sense.  We will refer to this as `pool` here.

Also, we want some way to set the "current branch" so that the pool
and "branch" is implied and the implied branch can be either a `commitish`
or a `ref`.

* branch `commitish` `sref` - create branch `sref` with parent commit object `commitish`
* branch -d `ref` - delete branch
* create `pool` - create the pool named pool
* delete `ref` `id` `id` ... - delete data objects from branch `ref` and update the branch to point to the new commit object
* drop `pool` - delete a data pool
* index `ref` `id` `id` ... - create index objects if they don't exist and record them in branch `ref`
* load `ref` `file` ... - load the files or stdin into data objects, create a commit object for the new data that points backward to the head of `ref`, and update branch `ref` to point to the new commit
* log `commitish`- display the commit object history starting at `commitish`
* ls - list pools in a lake
* ls `pool` - list branch names of a pool
* merge `commitish` `sref` - merge a commit history rooted at `commitish` into a branch `sref`
* query: `from commitish` - data scan

> * query: `from commitish:meta` branch or comit meta-scan (-no defaulting main here because thats pool-level meta)

Reviewer: ```suggestion
* query: `from commitish:meta` branch or commit meta-scan (-no defaulting main here because that's pool-level meta)
```
